common/trinary/trit_long.c

```diff
@@ -39,60 +46,72 @@ size_t long_to_trits(int64_t value, trit_t *out) {
-  size_t size, size_size;
-  size = nearest_greater_power_of_three(min_trits(value), power);
-  size_size = (*power + 1) / 2;
-  *start = size_size;
-  return size_size + size;
+size_t nearest_greater_multiple_of_three(size_t value) {
+  size_t rem = value % RADIX;
+  if (rem == 0) return value;
+  return value + RADIX - rem;
 }
 
 size_t encoded_length(int64_t value) {
-  size_t start, power;
-  return encoded_length_variables(value, &start, &power);
+  if (value == 0) return sizeof(encoded_zero) / sizeof(trit_t);
+  size_t length = nearest_greater_multiple_of_three(min_trits(llabs(value)));
+  // trits length + encoding length
+  return length + min_trits(pow(2, length / RADIX) - 1);
 }
 
-// trit_t *encode_long(int64_t value) {
-int encode_long(int64_t value, trit_t *out, size_t size) {
-  size_t size_size, power;
-  size_t end = encoded_length_variables(value, &size_size, &power);
-  if (size < end) {
-    goto encode_long_err;
+int encode_long(int64_t value, trit_t *trits, size_t size) {
+  if (size < encoded_length(value)) return -1;
+  if (value == 0) {
+    memcpy(trits, encoded_zero, encoded_length(0));
+    return 0;
+  }
+
+  size_t encoding = 0;
+  size_t index = 0;
+  size_t length = nearest_greater_multiple_of_three(min_trits(llabs(value)));
+
+  long_to_trits(value, trits);
+  for (size_t i = 0; i < length - RADIX; i += RADIX, index += 1) {
+    if (trits_to_long(&trits[i], RADIX) >= 0) {
+      encoding |= 1 << index;
+      for (size_t j = 0; j < RADIX; j++) trits[i + j] = -trits[i + j];
+    }
+  }
+  if (trits_to_long(&trits[length - RADIX], RADIX) <= 0) {
+    encoding |= 1 << index;
+    for (size_t i = 1; i < RADIX + 1; i++)
+      trits[length - i] = -trits[length - i];
   }
-  out[size_size - 1] = power & 1 ? -1 : 1;
-  long_to_trits(value, &out[size_size]);
+  long_to_trits(encoding, &trits[length]);
   return 0;
-encode_long_err:
-  return -1;
 }
 
-int64_t get_encoded_long(trit_t *trits, size_t length, size_t *end) {
-  size_t size = 1, i;
-  for (i = 0; i < length && trits[i] == 0; i++, size *= 3) {
-    size *= 3;
+int64_t decode_long(trit_t *trits, size_t length, size_t *end) {
+  if (memcmp(trits, encoded_zero, encoded_length(0)) == 0) {
+    *end = encoded_length(0);
+    return 0;
   }
-  if (trits[i] == -1) {
-    size *= 3;
+  int64_t value = 0;
+  size_t encoding_start = 0;
+  while (encoding_start < length &&
+         trits_to_long(&trits[encoding_start], RADIX) < 0)
+    encoding_start += RADIX;
+  if (encoding_start >= length) return -1;
+  encoding_start += RADIX;
+  size_t encoding_length = min_trits(pow(2, encoding_start / RADIX) - 1);
+  size_t encoding = trits_to_long(&trits[encoding_start], encoding_length);
+  for (size_t i = 0; i < encoding_start / RADIX; i += 1) {
+    int64_t tryte_value = trits_to_long(&trits[i * RADIX], RADIX);
+    if ((encoding >> i) & 1) tryte_value = -tryte_value;
+    value += pow(27, i) * tryte_value;
```

Comment:
Since variable `i` might only equal to limited possible numbers, would you consider to construct a dedicated lookup table to avoid duplicated computation caused by `pow`?

Reply:
Thanks, lookup table for this is actually ~18% faster than pow function !